presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSplitManager.java

```diff
@@ -229,7 +229,7 @@ public ConnectorSplitSource getSplits(
                 namenodeStats,
                 directoryLister,
                 executor,
-                splitLoaderConcurrency,
+                min(splitLoaderConcurrency, partitions.size() + 1), // Avoid over-committing split loader concurrency
```

Comment:
What if there are no partitions? Why +1?

Reply:
Unpartitioned tables still have a sentinel `"UNPARTITIONED"` entry here giving the list a size of 1. The `+1` piece here is because I measured a slight improvement with it compared to just partition size, presumably because the hand-off between the file iterator and partition queues and the transition to "no more splits" benefits from the extra waiting thread.

That said, my testing was done a much older version of presto. Looking through the current `BackgroundHiveSplitLoader` implementation it appears that it's actually possible to have a greater concurrency than 1 per partition when multiple read paths are present, so that definitely brings the validity of this approach into question.

Reply:
Atleast in non-ACID flow, doesn't it suffice to have 1 thread per partition? the HiveFileIterator is created per partition path.

Reply:
@rohangarg looks like you're right- only transactional tables can potentially benefit from more than 1 thread per partition as implemented today. I've updated the PR to use one thread per partition for non-transactional tables.

Reply:
> Unpartitioned tables still have a sentinel "UNPARTITIONED" entry here giving the list a size of 1

I'm pretty sure for unpartitioned tables we might want concurrency higher that 2. Imaging multi TB table with lots of partitions. Also bucketed tables are also unpartitioned, right?

Reply:
Unpartitioned tables can only benefit from a single thread (as implemented today) because of the nature of the iterators they use. The same goes for buckets as far as I can tell although maybe I’m missing something in this implementation. Today, submitting more than one task per partition just increases the number of threads contending for the lock and does not result in any increase in the rate of work being done.

Reply:
> Today, submitting more than one task per partition just increases the number of threads contending for the lock

Is the lock contention an issue when other threads do not have work to do? In practice only as many threads do processing as there is work and extra threads should not interfere too much, right?
Do you observe some penalty when there are too many threads?

Reply:
The extra threads that perform no work cause two problems:
- they create extra lock contention and will typically fall through to the path that acquires the split loader write lock, locking out tasks that can do work from acquiring the read lock for as long as it takes for all other tasks to release their read lock 
- they take up threads in the cached threadpool executor which could have been used for other table scans or might never have needed to be created at all.

The second part is the largest general improvement since threads are a heavy resource to create when they’re not need, but there are worst case scenario’s around the first problem that can stall split loading for long periods of time.